src/cmdstan/command_helper.hpp

```diff
@@ -0,0 +1,589 @@
+      ++idx;
+    }
+  }
+  return params_r_ind;
+}
+
+/**
+ * Extract and validate constrained parameter inputs.
+ * Input file contains definitions for all model parmeters.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @param model Stan model
+ * @return Eigen vector of parameter estimates.
+ */
+std::vector<std::vector<double>> get_cparams_r(
+    const std::string &fname, const stan::model::model_base &model) {
+  std::vector<std::string> param_names;
+  std::vector<std::vector<size_t>> param_dimss;
+  stan::services::get_model_parameters(model, param_names, param_dimss);
+
+  size_t num_upars = model.num_params_r();
+  std::vector<double> params(num_upars);
+  std::vector<std::vector<double>> params_r_ind = {params};
+  std::vector<int> dummy_params_i;
+  std::shared_ptr<stan::io::var_context> cpars_context = get_var_context(fname);
+  std::vector<std::string> input_cpar_names;
+  cpars_context->names_r(input_cpar_names);
+  std::stringstream msg;
+  for (const std::string &m_param_name : param_names) {
+    if (!cpars_context->contains_r(m_param_name)) {
+      msg << "Constrained value(s) for parameter " << m_param_name
+          << " not found!";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+  model.transform_inits((*cpars_context), dummy_params_i, params_r_ind[0],
+                        &msg);
+  return params_r_ind;
+}
+
+/**
+ * Read fitted params from sample in StanCSV format.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @param model Stan model
+ * @return Eigen vector of parameter estimates.
+ */
+std::vector<std::vector<double>> get_cparams_r_csv(
+    const std::string &fname, const stan::model::model_base &model) {
+  stan::io::stan_csv fitted_params;
+  size_t col_offset, num_rows, num_cols;
+  get_fitted_params(fname, model, fitted_params, col_offset, num_rows,
+                    num_cols);
+  std::vector<std::vector<double>> result(num_rows);
+  for (size_t i = 0; i < num_rows; ++i)
+    result[i] = stan::math::to_array_1d(
+        fitted_params.samples.block(i, 0, 1, num_cols));
+  return result;
+}
+
+/**
+ * Given a set of parameter values, call model's log_prob_grad
+ * method and send output to a CSV file.
+ *
+ * @param model Stan model
+ * @param jacobian jacobian adjustment flag
+ * @param params_set array of unconstrained parameter values
+ * @
+ */
+void services_log_prob_grad(const stan::model::model_base &model, bool jacobian,
+                            std::vector<std::vector<double>> &params_set,
+                            int sig_figs, std::ostream &output_stream) {
+  // header row
+  output_stream << std::setprecision(sig_figs) << "lp__,";
+  std::vector<std::string> p_names;
+  model.constrained_param_names(p_names, false, false);
+  for (size_t i = 0; i < p_names.size(); ++i) {
+    output_stream << "g_" << p_names[i];
+    if (i == p_names.size() - 1)
+      output_stream << "\n";
+    else
+      output_stream << ",";
+  }
```

Comment:
This is actually a bug in the released version of this feature, but this needs to be `unconstrained_param_names`. 

Try it with a simplex parameter and you'll get a misformed CSV result:

```csv
lp__,g_theta.1,g_theta.2
-1.20397,-0.4
```

Reply:
corrected, plus unit test to check.